torch/csrc/api/include/torch/nn/pimpl-inl.h

```diff
@@ -45,3 +45,28 @@ struct is_module_holder_of : is_module_holder_of_impl<
                                  is_module_holder<T>::value,
                                  decay_t<T>,
                                  decay_t<C>> {};
+
+// A collection of templates that allow deducing the return type of the
+// `forward()` method, but only if a module actually has a `forward()` method,
+// and otherwise deduces to the type `void`.
```

Comment:
@smessmer would you mind reviewing these templates?

Reply:
`has_forward` has a quite complex implementation that maybe can be simplified with `void_t` (it's a c++17 feature but we have `c10::guts::void_t` for C++11), but that's actually not in this diff but already landed. The templates added here look good.